keps/0036-kudo-operator-toggle-behavior.md

```diff
@@ -0,0 +1,147 @@
+creation-date: 2021-04-30
+last-updated: 2021-04-30
+status: provisional
+see-also:
+  - KEP-23
+  - KEP-29
+---
+
+# Parameter-controlled toggling of KudoOperator tasks
+
+## Table of Contents
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+  * [Interface](#interface)
+     * [Validation](#validation)
+     * [Changing enablingParameter on upgrade](#changing-enablingparameter-on-upgrade)
+  * [Implementation Notes](#implementation-notes)
+  * [Risks and Mitigations](#risks-and-mitigations)
+* [Drawbacks](#drawbacks)
+* [Alternatives](#alternatives)
+
+<!-- [Tools for generating]: https://github.com/ekalinin/github-markdown-toc -->
+
+## Summary
+
+This KEP aims to add an ability to switch an operator dependency on and off
+via a specified parameter of a parent operator. Namely, as a result of
+implementing this KEP, `KudoOperator` task definition will get an optional field
+containing a name of a boolean parameter (of the parent operator), the value
+of which will determine whether the child operator instance should exist.
+
+## Motivation
+From [#1775](https://github.com/kudobuilder/kudo/issues/1775):
+> Currently, KudoOperator tasks don't support toggle behavior which
+> significantly complicates the handling of the optional dependencies. A good
+> example could be Spark and History Server or Kafka and Schema Registry. The
+> only possible options to handle it today are either include the templates
+> as-is into the parent operator and control the contents at the template level
+> or to implement a job (via Toggle task) which deploys/removes the dependency
+> based on a parameter value. Also, cascading updates/upgrades become
+> complicated in these scenarios.
+
+### Goals
+
+* Add an ability to enforce existence/**non-existence** of the child
+  operator instance (and all its children) via a parent operator parameter,
+  i.e. to toggle the dependency.
+
+* Make parent operators specifying an invalid parameter for toggling fail
+  instance admission (as opposed to simply failing a plan with misconfigured
+  `KudoOperator` task).
+
+### Non-Goals
+
+* Chain validation of child instances. `KudoOperator` configurations that result
+  in failed instance admission of a child do not fail parent instance admission,
+  only plan execution. Functionality introduced by this proposal is also
+  affected by this: for example, if a parent switches on a child that uses a
+  non-existent parameter for toggling a grandchild, parent instance admission
+  will succeed, only the corresponding parent plan will fail.
+
+* OperatorVersion validation. This proposal adds one more way to create
+  an operator that just cannot be installed regardless of instance parameters.
+
+* Introducing an `UninstallKudoOperator` task. An ability to unconditionally
+  remove a child operator (for example, installed by a previous version), while
+  no more difficult, is not a goal of this proposal.
+
```

Comment:
before getting to "proposal"... we need more information outlined to set expectations..
For examples questions / comments below, assuming Spark operator dependent on History Server and the KUDO spark operator is configured for a dependent KUDO History Server.

When KUDO Spark is installed... it will install KUDO History Server.  Resulting in KUDO having "ownership" as a controller over the history server.  Resulting in param changes for a kudo update being managed for the history server.

Questions:
1. IF we opt-out of the dependency, does KUDO own it or not own it?  It seems it would not but the KEP should make that clear... with all the responsibilities for the dependency being removed from kudo.  (like updates, deletes etc.)
2. IF we opt-out, (and I'm going off really old memory here)..  I can't remember if we have the ability to have output values from a dependency... for instance a dependent zookeeper, provide a zookeeper client details to a parent.. if we have that, how does that work for a parent that opts out?
3. If we opt-out, what happens for `kudo update` for parameters which are for an opt-out dependency?  Do we error out?  does it ignore it?
4. same for a `kudo upgrade`
5. for an opt-out, can an `kudo update` reverse the decision?   Opting in and provisioning a new dependency?
6. What does the plan status look like for an opt-out... now it shows a tree of phases including all dependencies... does it show the dependencies?  Does it reflect the opt-out?

There is likely more... we need these questions answered in the KEP to frame a mental model for discussions, expectations and awareness.  A feature of this magnitude requires some thoughts around impact to all other features and usability.    It really is required prior to any proposal as the proposal will be evaluated against all the edge cases.

We don't have to boil the ocean..  this KEP is "provisional"... the first go of it, doesn't require a proposal... just a lot of detail of all the edge cases and a strong case around the acceptance criteria.  It is also reasonable to identify cases which are too complex and will not be supported but they need to called out in this doc which becomes the foundation by which we add the feature.

Reply:
@kensipe Thanks for these questions!

Regarding (1), I've tried to reword this KEP to make it clear that the goal is not toggling of a dependency relationship between two operator instances, but controlling whether the dependency instance exists by means of a parameter.

(2) is not an issue: as of today, there are no means for obtaining an output from the dependency. 
I've tried to address (3) - (6) are in the Non-Goals section.

Please take a new look; hopefully, this makes more sense now.

Reply:
@asekretenko  sorry for delays.. I will make time to review again tomorrow morning... it should be there by your mid-day.